### ISSUE TITLE
Exit with code 1 due to network error: ContentNotFoundError

### ISSUE THREAD

**Issue by djvs** [wkhtmltopdf/wkhtmltopdf#2051]
With v0.12.1, PDFKit wrapper and  --load-error-handling ignore:

Exit with code 1 due to network error: ContentNotFoundError
command failed (exitstatus=1): /usr/local/bin/wkhtmltopdf --quiet --page-size Letter --margin-top 0.75in --margin-right 0.75in --margin-bottom 0.75in --margin-left 0.75in --encoding UTF-8 --load-error-handling ignore - /tmp/5463d01c6f706554720b0100.pdf

**Comment by TheShrike** [comment]
Same issue, I get the error when the html request some resources on a domain managed by the same server, solved adding a new row in the hosts file of the server:

127.0.0.1 domain_used

**Comment by Joe90k** [comment]
@TheShrike I have tried your solution as I am having the same problem. All content renders fine as long as its not on the same server that wkhtmltopdf is running on. I have added the domain to my hosts file in CentOS 6 and saved but I get the same result. Is there anymore info or help you can give?

**Comment by TheShrike** [comment]
@Joe90k Are you sure the new hosts file is used? If you try to ping your domain you should get this:
PING your_domain (127.0.0.1) etc.

**Comment by Joe90k** [comment]
@TheShrike yes I do, wkhtmltopdf still generates a redirect page which is the default page for any domain that is pointing at our server but has not been assigned a web root. 

I wish there was a way of seeing where/how wkhtmltopdf looks for the domain!

**Comment by ashkulz** [comment]
What happens if you `wget` the URL you are trying to convert on the server itself? If it generates the redirect page, then that is a server setup issue.

**Comment by Joe90k** [comment]
@ashkulz Simple, hadn't thought of that. It does indeed generate the redirect page. It seems there must be something amiss with the server config. Thanks for the help.

**Comment by vannderr** [comment]
I had this same ContentNotFoundError with a simple wkhtmltopdf command.
Turned out that my html page had a <script> line in the header pointing to a .js file, with a relative url, something like '../blabla.js'
Looks like wkhtmltopdf cannot handle that.
So I took out that line - and the error disappeared :-)

**Comment by gkertai** [comment]
We've also run into this problem on our development servers, and as it turns out, the [Symfony profiler](http://symfony.com/doc/current/book/internals.html#profiler) injected its debugging toolbar into the pages we were trying to render, which made wkhtmltopdf choke. The JavaScript of the toolbar must have changed with a recent Symfony update, making this error pop out of nowhere. Disabling the profiler in the affected controllers solved it.

**Comment by pkpp1233** [comment]
Is there a general solution in cases where you don't have control over the relative URLs? Potentially a flag to place that lets wkhtmltopdf ignore the error?

**Comment by chreds** [comment]
I'm running into the same thing. I've tried with "load-error-handling" as both "ignore" and "skip". Neither seems to ignore the content load errors for all the files that are failing. The output show it's skipping some of them but not all?

**Comment by kenorb** [comment]
The same here:

    http://stackoverflow.hewgill.com/questions/282/357/1.html

**Comment by mtalcott** [comment]
I'm seeing that slight changes in the URL of certain resources completely changes the load error handling behavior of wkhtmltopdf. Additionally, `--load-error-handling` has no effect on such behavior.

Here are some test cases with results: https://gist.github.com/mtalcott/447c982e59bcdc91b25d

**Comment by mvl22** [comment]
I've run into this bug also.

We are generating some HTML and using the ```<base href="http://site.example.com">``` tag to resolve URLs beginning with / to a full URL.

However, if a CSS reference like
```<link rel="stylesheet" type="text/css" href="/path/to/mainsite.css?13452" />```
is present, that causes the problem discussed above.

Changing this to remove the versioning number, i.e.
```<link rel="stylesheet" type="text/css" href="/path/to/mainsite.css" />```
causes the program to complete cleanly with a success code.

The file has similar query strings with javascript files and general URL links, but they do not trigger the problem. Removing just the query string in the CSS reference avoids the bug being triggered.

**Comment by luzfcb** [comment]
This issue bite me too.
In my case, i have the follow code in my template 

```
    <script>if (!window.jQuery) {
        document.write('<script src="/static/core/js/jquery/1.11.3/jquery.min.js"><\/script>');
        document.write('<script src="/static/core/js/jquery/1.11.3/jquery-migrate.min.js"><\/script>');
    }
    </script>
```
after removing this code, all goes as expected

**Comment by c4bo3l** [comment]
I have solved this issue by put full path of your js and css files. It will be like 
`<link rel="stylesheet" href="http://blabla.com/css/cssFilename.css"/>`
`< script type="text/javascript" src="http://blabla.com/js/jsFileName.js"/></script>`

**Comment by joshribakoff** [comment]
Isn't there a separate bug in that this error message sucks? I'm running across this intermittently in our logs, yet whenever I debug I cannot replicate it. I have no way to know which script is intermittently failing because the error message doesn't give that info. This should be fixed, the bad error message is a bug.

**Comment by hariprasadkulkarni** [comment]
I'm facing a similar issue. My html has no request to an external resource. The file is created, but I get this error.

    Loading pages (1/6)
    Warning: Failed loading page http:// (ignored)
    Counting pages (2/6)
    Resolving links (4/6)
    Loading headers and footers (5/6)
    Printing pages (6/6)
    Done
    Exit with code 1 due to network error: HostNotFoundError

And the command I'm using is this,
cat test.html | wkhtmltopdf --margin-bottom "1" --margin-left "1" --margin-right "1"
--margin-top "5" --orientation "Landscape" --page-size "Legal"
--load-error-handling "ignore" --load-media-error handling "ignore"
--disable-smart-shrinking "" - "out.pdf"

This is particularly problematic when I'm using the JS library. On getting this error, the error handler in the callback function is triggered, while the file is generated.

I'm using wkhtmltopdf 0.12.2.1 (with patched qt) on Ubuntu 14.04.

**Comment by jeroendesloovere** [comment]
Same bug over here when trying to pdf a locally URL and I'm even using 'abolute paths' everywhere and no JavaScript.
When giving an external URL everything works as expected.

**Comment by brandonros** [comment]
I just got this error (despite passing both `--load-error-handling "ignore"` and `--load-media-error-handling "ignore"`.

Is it possible that Webkit was updated recently? I don't know how my server would have gotten it, but because I have `<img src="REPLACE_ME">` somewhere in a JSON object, the rendering engine seems to attempt to preload the resource despite it not existing?

**Comment by joshribakoff** [comment]
@brandonros not quite sure that makes sense. Somewhere in a JSON object? So you have a json (string), or an object (not json)? "Where" is this value? stored in a javascript variable? or is the value written to the DOM. The browser will try to load any `src` tags though. That's why frameworks like angular have `ng-src`. Try renaming `src` to `brandon-src`, and then only write the URL into `src` with the real value rather than `REPLACE_ME`.

**Comment by jeff-h** [comment]
Adding the `<base>` tag, as per https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2051#issuecomment-137267732 above, fixed this problem for me, despite my Drupal-generated page having query strings on all the CSS files (in contradiction to the latter part of the same comment).

**Comment by Garric15** [comment]
Solution here : https://github.com/wkhtmltopdf/wkhtmltopdf/issues/1634#issuecomment-40597921

> I simply had to change:
> `<script src="//cdnjs.cloudflare.com/ajax/libs/numeral.js/1.4.5/numeral.min.js"></script>`
> 
> to:
> `<script src="http://cdnjs.cloudflare.com/ajax/libs/numeral.js/1.4.5/numeral.min.js"></script>`
> 
> (added the HTTP).

**Comment by HorizonShadow** [comment]
Seems you also get this error if you have relative paths in your CSS.

For example, this styling from semantic-ui will cause this error:

    @font-face {
      font-family: 'Icons';
      src: url("themes/default/assets/fonts/icons.eot");
      src: url("themes/default/assets/fonts/icons.eot?#iefix") format('embedded-opentype'), url("themes/default/assets/fonts/icons.woff2") format('woff2'), url("themes/default/assets/fonts/icons.woff") format('woff'), url("themes/default/assets/fonts/icons.ttf")     format('truetype'), url("themes/default/assets/fonts/icons.svg#icons") format('svg');
      font-style: normal;
      font-weight: normal;
      font-variant: normal;
      text-decoration: inherit;
      text-transform: none;
    }

**Comment by ztane** [comment]
WT... this started happening to me right now, nothing has changed that I know and I think I've always had relative links. I am using the prebuilt binary as well...

**Comment by ztane** [comment]
Ah nvm, it was a hosts issue (I had suppressed fonts.google.com but didn't remember it). However the "ignore" options described above didn't work

**Comment by wimvds** [comment]
I just did some debugging because I have the same issue with Google fonts imports.

The HTML contains the following in the header (amongst other things) :
`<style>
@import url(https://fonts.googleapis.com/css?family=Roboto:400,500,700,300);
</style>
`

When I try to create the PDF without an internet connection this always fails with 'Exit with code 1 due to network error: HostNotFoundError', even though I used --load-error-handling 'ignore' and --load-media-error-handling 'ignore'. If I comment the imports in the template everything works as expected, and the PDF gets generated.

Hope this helps in finding the root cause.

**Comment by benf-wspdigital** [comment]
As joshribakoff says, the message makes this far more difficult to debug than it could be. That is a separate issue, #2228.

**Comment by ztane** [comment]
Yeah, this still is a serious bug. I do not really give a s*t about not having some fonts there. I just want my output.

**Comment by jeroendesloovere** [comment]
This bug has made me switch to another PDF generator. Which is what nobody wants...

**Comment by djvs** [comment]
I ended up manually downloading assets to get around this bug.   Very ugly
but a hack you can use if badly needed.

**Comment by Salmon-Bard** [comment]
I had this issue too. Checked various option and found the root of this problem and most of the issues reported here on various network and content not found error.

The issue is mainly due to a bug in qtwebkit qt4 patched version source code used in wkhtmltopdf. qtwebkit is not able to correctly resolve browser dns redirection and multi dns calls to resource located on one or more resources within the page.  An upgrade to newer qtwebkit Qt5 and a revised batch will likely fix all theses bugs.

Until maintainers have time to work on this or someone with the expertise lends a hand (sadly not me of course :- ( )  a workaround I found fixes the issue.

 calling your page using server IP without domain names fixed the problem. so instead of using `http://mydoamin.com/page_url  `

in your wkhtmltopdf argument call
` http://xx.xx.xx.xx/page_url `

And if you have linked files in your webpage not loading try calling them through server IP and it will likely work.

hope you find this helpful.

**Comment by ztane** [comment]
Yeah but this still does not explain *why* by default one 404 needs to *abort* the whole rendering process. There ought to be a way to configure it.

**Comment by Salmon-Bard** [comment]
There is away to configure behavior on 404 but it does not work, the **--ignore** flag should allow bypassing unloaded content, unfortunately it does not!!!

**Comment by emanuelecasadio** [comment]
I am now getting something like one thousand "code 1 returns" on a daily basis, despite running with the following params `['--quiet','--load-error-handling','ignore','--stop-slow-scripts','--width','1920','--height','1080','--format','jpeg',uri,'-']`.

This seriously needs to be fixed.

**Comment by etagwerker** [comment]
I ran into this issue as well. I tried with these flags: `--load-error-handling ignore` and  `--load-media-error-handling ignore` but the error kept happening. In my case, the content that wasn't found were two images. 

I ended up `--no-images` which is not ideal.

**Comment by emanuelecasadio** [comment]
I am experiencing exactly the same problem.

**Comment by utkarsh27a** [comment]
Hey guys, I faced the same issue today in laravel.
_Exit with code 1 due to network error: ContentNotFoundError_
**I tried to wrap my all css and script link with url() function.**
example: 
url('/css/bootstrap.min.css')
url('/js/jquery-2.1.1.js') 

and my code working. Now every thing is fine. :)

**Comment by ChiggerChug** [comment]
I get that this may not be fixed in the short term, but it there a legit way to suppress this specific error? I know certain things wont load in my page but I need the output without wkhtmltopdf crashing my node server because it exited with an error code.

**Comment by wsitko** [comment]
While reading one of the answers above I realized that the output image file is generated despite the fact that the error code is returned.
I used to verify result by testing exit code, but I switched my code to check if the output file exists.
For sure it's not a best practice, but it works brilliant as long as server errors can't be suppressed.

**Comment by waspinator** [comment]
I'm using WKHTMLTOPDF with cakephp (cakepdf), and I'm able to generate a pdf successfully once every 1-2 minutes. the input/output is a blank page, and without changing anything and running it more often than every 1-2 minutes generates this error.

Using 0.12.3 on ubuntu 16.04 x64

**Comment by MalakDudhia** [comment]
Hello All,

Have you check Special character in link or not?If there is special character than you should encode it .Only encode special character not any other.

**Comment by ninapavlich** [comment]
Same error here.

**Comment by cgaugel** [comment]
Any news? Still the same error here!

**Comment by hellvix** [comment]
I don't think this will be fixed so soon, given how long it has been opened. But the issue is due to a relative path. You have to include the full file path. It might be occurring with your JavaScript file. Instead of using `/path/to/file` use `http://domain.com/path/to/file`.

**Comment by benf-wspdigital** [comment]
@hellvix: Both your examples use absolute path. You've described a difference between a absolute filesystem path (`/path/to/file`) and a URL containing an absolute path (`http://domain.com/path/to/file`).

Are you saying we should expect the program to fail when specifying an absolute path of a file? That only a URL will work?

**Comment by hellvix** [comment]
@benf-wspdigital That is not the absolute system file location, it is path for the file file in the URL after the domain. Basically the difference between `HttpRequest.get_full_path()` and `HttpRequest.build_absolute_uri` described in [here](https://docs.djangoproject.com/en/1.10/ref/request-response/#methods).

If you hard code your HTML (that will be transformed in PDF) the full URL of the file (including the domain), it will work. Instead of using `{% static 'path/to/file' %}` use `http://domæne.com/path/to/file`. A way to know which resource file is missing, is going to `/tmp/` (assuming you are on Linux) and check the temporary files wkhtmltopdf generates and tries to transform into PDF. If you open it with your browser and inspect the source of the page, you'll see the missing resources there.

I am struggling with this issue for soem time. The problem is that I have a lot of resources that are dynamically generated. I wanted this PDF page to be as generic as possible without much patching, but this URL thing is giving me a huge head ache.

**Comment by benf-wspdigital** [comment]
> If you hard code your HTML (that will be transformed in PDF) with the full URL of the file (including the domain), it will work.

Thanks, that's what I thought you meant.

I don't think that counts as a work-around; what's needed is to specify static files on the local filesystem.

**Comment by ajaxray** [comment]
I had this issue several times and the root cause was same - the renderer could not load a/some javascript/image/font file. After resolving missing resource, the issue was fixed.
BTW, to quickly find the missing resource you may open the page URL from the browser with keeping `console > Network` panel open.

**Comment by omrqs** [comment]
This error occurs when url are defined as //url... in place of https://url....
Keep the protocol. 
wkhtmltopdf try find double slash as local path...

Solved for me.

**Comment by meghajhala** [comment]
I have added these flags based on the suggestions above: `--load-error-handling ignore and --load-media-error-handling ignore`

And I am using the version: wkhtmltopdf 0.12.1 (with patched qt).

However still getting this error for an HTML page with video:
**Exit with code 1 due to network error: ContentNotFoundError**

Please suggest and any help on this will be appreciated.

**Comment by Laubeee** [comment]
For those having issues with fonts ALTHOUGH they are available locally (which fixed the issue in previous versions such as 0.9.6) - it seems there is still a problem with the relative path in font-face. So if you use the absolute path (with e.g. with $_SERVER['DOCUMENT_ROOT']) it now works for me!

**Comment by bismaya123** [comment]
Hi Everyone, i need a small help from you guys. here is command which was sent to server.

--zoom 0.75 --page-size Letter --print-media-type --load-media-error-handling ignore --footer-left "Thomson Reuters | MICROSOFT CORP Company Report" --footer-right [page] --javascript-delay 100000 --load-error-handling ignore toc "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#CoverPage/cc957ef8053c46aeabe42250c81bbeee?isShowProfileName=true&isShowDescription=false&isShowCompanies=true&isShowProfile=true&isShowPracticeArea=true&isShowDockets=true&notesText=&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#MonitorCourt/cc957ef8053c46aeabe42250c81bbeee?monthsInterval=12&componentDetail=false&includeLegend=true&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#Representation/cc957ef8053c46aeabe42250c81bbeee?graphType=TotalEvents&firm=&listType=TopTen&componentDetail=false&pai=&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#AttorneyRepresentation/cc957ef8053c46aeabe42250c81bbeee?displayType=all&monthsInterval=12&componentDetail=false&pai=&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#CompanyComparisonV2/cc957ef8053c46aeabe42250c81bbeee?monthsInterval=12&componentDetail=false&pai=&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#RecentDockets/cc957ef8053c46aeabe42250c81bbeee?componentDetail=false&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#RelatedNews/cc957ef8053c46aeabe42250c81bbeee?cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#Summary/cc957ef8053c46aeabe42250c81bbeee?marketOverview=true&netIncome=true&incomeStatement=true&netWorth=true&balanceSheet=true&listView=true&biographies=false&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#OfficersAndDirectors/cc957ef8053c46aeabe42250c81bbeee?marketOverview=true&netIncome=true&incomeStatement=true&netWorth=true&balanceSheet=true&listView=true&biographies=false&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#IncomeChartStatement/cc957ef8053c46aeabe42250c81bbeee?marketOverview=true&netIncome=true&incomeStatement=true&netWorth=true&balanceSheet=true&listView=true&biographies=false&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" "https://intelligencecenter.qed.thomsonreuters.com/Clients/Widget/#NetWorthChartStatement/cc957ef8053c46aeabe42250c81bbeee?marketOverview=true&netIncome=true&incomeStatement=true&netWorth=true&balanceSheet=true&listView=true&biographies=false&cto=-330&printMode=true&reportName=AlexTest_062017.pdf" /home/asadmin/test.pdf

here i'm getting sometimes pdf without content except header footer and message like "Your browser is not optimized for Monitor SuiteIntelligence Center." and some times it failed with following error

"An error happened while creating the pdf. please try again later" 

Though its an intermittent issue, but why i'm getting this error.

**Comment by ragnarkurmwunder** [comment]
I got same error message "Exit with code 1 due to network error: ContentNotFoundError".
The problem was that source HTML referenced resources which had only path int them. For example:
```html
<svg role="presentation">
<use xlink:href="/svg/sprite.svg" xmlns:xlink="http://www.w3.org/1999/xlink" />
</svg>
```

**Comment by to-kra** [comment]
any update here ?

**Comment by hellvix** [comment]
It won't be fixed. Instead, read what I posted above. There is a work-around for this problem.

> I don't think this will be fixed so soon, given how long it has been opened. But the issue is due to a relative path. You have to include the full file path. It might be occurring with your JavaScript/CSS files. Instead of using /path/to/file use http://domain.com/path/to/file.

**Comment by to-kra** [comment]
@hellvix the problem is that we cannot fix source content. why you think its not gonna be fixed. i think many people using wkhtmltopdf and having this issue.

**Comment by hellvix** [comment]
@to-kra Yeah, I know we can't edit the source and that's why I am saying it won't be fixed. Look when the issue in this topic was originally reported. It has been opened for years.

**Comment by to-kra** [comment]
@hellvix who is then responsible for fixing wkhtmltopdf ? :(

**Comment by hellvix** [comment]
@to-kra I would guess the guy owning this project on GitHub.

**Comment by mvl22** [comment]
> who is then responsible for fixing wkhtmltopdf

It's an open source, community project. If no-one kindly volunteers their time, or finds funds to have someone to pay to fix it, then it won't get fixed.

**Comment by montao** [comment]
I have this issue too just trying to render a pdf from html. It really sucks. Sorry for being negative but this defect really makes wkhtmltopdf rather unusable.

**Comment by akalongman** [comment]
I found a workaround:
```php
// Replace all ?sadasd query strings from static file includes
$html = preg_replace('#(\.css|\.js)\?[^"]+#',  '$1', $html);
```
before passing html to `wkhtmltopdf`

**Comment by Arkanius** [comment]
In case of Ubuntu, try to update your `libssl` with: 

```sudo apt install libssl-dev=1.0.2g-1ubuntu4.10```

**Comment by quantum-x** [comment]
## You too?
For anyone that has stumbled onto this thread via a search engine, dismayed that the issue was 'closed without reason', read on.

### This is a generic error, linked to specific cases
This is a generic error that wkhtmltopdf generates when **it cannot find a specific resource**.
This means that **your** html / code passed to wkhtmltopdf contains links to resources that wkhtmltopdf **cannot find**.

**The reasons it cannot find them are several and varied.**

### The most simple:
- Perhaps you made a typo in the resource
- Perhaps the resource doesn't actually exist

An easy way to check if you fall into this category, `tail` your web server logs.
You may notice 404 / 403 errors being generated, ie
```
Request URL: /bower_components/fontawesome/webfonts/fa-regular-400.svg
Result: 404
```

### The more complicated
If you're using absolute URLs for your content, there could be other reasons:
- Your webserver can't reach itself
- You have multiple web-servers and the resource in question isn't on all of them
- Your server has outdated ca-certs / ssl libraries

**Solution**
Go through the source passed to your PDF, and check every single absolute URL from the CLI of the machine that you're attempting to generate the PDF, ie:
```bash
curl -s -o /dev/null -w "%{http_code}" http://your.com/url/path
```

If it doesn't return '200' - this may be where your problem lies, and will merit further investigation.

Or, you may be incredibly unlucky and have actually discovered a bug - lodge an issue and good luck!

**Comment by MichaelKim0407** [comment]
Sometimes wkhtmltopdf will exit 1 with 'ContentNotFoundError' even if the PDF generation is finished.
It will display 'Done' at the second last line of the output.
In this case we can actually use the generated content.

If you are using the `pdfkit` Python package like me, you are welcome to use my fix before a patch is released: https://github.com/MichaelKim0407/python-pdfkit
Pull request for `pdfkit`: https://github.com/JazzCore/python-pdfkit/pull/107

**Comment by rami7yakoub** [comment]
For me, the problem was that one of the "src" tags in the HTML file was pointing to a wrong directory (because forgot to change it after switching to another machine)

**Comment by yangweijie** [comment]
> Same error here. Happening with an image.
> Tried with IP versus Hostname, didn't change anything 😢

i transform src to base_64 format for img

**Comment by lucenarenato** [comment]
Does anyone have the solution now in 2019?
```sh
wkhtmltopdf --lowquality --header-html '/tmp/knp_snappy5d5d2fce6625d0.53804995.htm
> l' '/tmp/knp_snappy5d5d2fce660806.00715860.html' '/tmp/knp_snappy5d5d2fce661d57.43165999.pdf'
```
Loading pages (1/6)
Error: Failed loading page http:// (sometimes it will work just to ignore this error with --load-error-handling ignore)
Exit with code 1 due to network error: HostNotFoundError

**Comment by lucenarenato** [comment]
Achei a Solução: https://github.com/lucenarenato/docker-alpine-wkhtmltopdf
Solved: Had to use a version of wkhtmltopdf that is patched for Alpine. github.com/madnight/docker-alpine-wkhtmltopdf

**Comment by lucenarenato** [comment]
RUN apk add --update --no-cache \
    libgcc libstdc++ libx11 glib libxrender libxext libintl \
    ttf-dejavu ttf-droid ttf-freefont ttf-liberation ttf-ubuntu-font-family
# binary is static compiled against libssl1.0 libcrypto1.0 see issue #3
RUN echo 'http://dl-cdn.alpinelinux.org/alpine/v3.8/main' >> \
    /etc/apk/repositories && \
    apk --update --no-cache add libssl1.0 libcrypto1.0
# on alpine static compiled patched qt headless wkhtmltopdf (47.2 MB)
# compilation takes 4 hours on EC2 m1.large in 2016 thats why binary
COPY wkhtmltopdf /bin

**Comment by digitalfox** [comment]
I solved it... by switching to weasyprint....
https://github.com/Kozea/WeasyPrint/

**Comment by vikashsrivastava1111989** [comment]
ss

**Comment by rainyman2012** [comment]
I had this problem before. and solve that like below.

![wkhtmltopdf](https://user-images.githubusercontent.com/9016471/82428191-66006180-9a9f-11ea-8314-bcc16379425d.png)

In the above example i had some "src" files and "url"s  that they refer to the static directory but the static directory didn't exist so wkhtmltopdf thrown me that error. for example:

src: url("file:///home/ehsan/Projects/Example/main/sib/**static**/WebYekan.eot");

and one more important thing i have to say is all the file paths in html files have to be an absolute path. do not use relative path at all. 

I hope this help you.

**Comment by ABHISHEK-KEDAR-21** [comment]
```
OSError: wkhtmltoimage reported an error: 
Loading page (1/2) 
[> ] 0% 
[======> ] 10% 
[==========================> ] 44% 
Warning: Failed to load file:///assets/frappe/css/bootstrap.css (ignore) 
Warning: Failed to load file:///assets/frappe/css/font-awesome.css (ignore) 

Error: Failed to load file:///assets/frappe/css/fonts/fontawesome/fontawesome-webfont.woff?v=4.7.0, with network status code 203 and http status code 0 - 
Error opening /assets/frappe/css/fonts/fontawesome/fontawesome-webfont.woff: No such file or directory 

[============================================================] 100% 

Error: Failed to load file:///assets/frappe/css/fonts/fontawesome/fontawesome-webfont.ttf?v=4.7.0, with network status code 203 and http status code 0 - 
Error opening /assets/frappe/css/fonts/fontawesome/fontawesome-webfont.ttf: No such file or directory 

Error: Failed to load file:///assets/frappe/css/fonts/fontawesome/fontawesome-webfont.svg?v=4.7.0#fontawesomeregular, with network status code 203 and http status code 0 - 
Error opening /assets/frappe/css/fonts/fontawesome/fontawesome-webfont.svg: No such file or directory 

Rendering (2/2) 
[> ] 0% 
[===============> ] 25% 

Warning: Received createRequest signal on a disposed ResourceObject's NetworkAccessManager. This might be an indication of an iframe taking too long to load.
 [============================================================] 100% Done 

Exit with code 1 due to network error: ContentNotFoundError
```

In my case, I am getting above error, I think this is related with the query string ```v=4.7.0```

Any pointers please?

**Comment by Zswdhy** [comment]
wkhtmltopdf  index.html 7.pdf

Loading pages (1/6)
Warning: Blocked access to file E:/David/company/projects/html/cart/first_test.2/static/js/jquery.js
Warning: Blocked access to file E:/David/company/projects/html/cart/first_test.2/static/js/vue.js
Warning: Blocked access to file E:/David/company/projects/html/cart/first_test.2/static/js/first_part.js
Warning: Blocked access to file E:/David/company/projects/html/cart/first_test.2/static/css/index.css
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Warning: Blocked access to file E:/David/company/projects/html/cart/first_test.2/images/logo.png
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Counting pages (2/6)
Resolving links (4/6)
Loading headers and footers (5/6)
Printing pages (6/6)
Done
Exit with code 1 due to network error: ProtocolUnknownError

**Comment by ThorthChariot** [comment]
Hey.
I found a solution for this. i had the same problem and tried different many versions of wkhtmltopdf. 
I have a server with nginx + others packages.
Right now for me is working even the last version, 100%.

The solution is super simple, just go to your host file on server and add there the ip of the machine you running also the domain (  127.0.0.1  example.com ), save the host file and job done. 

Is the same problem when you try to run a curl on a page o the same server. the call reach the server, and return a 401 code, since he have problem to identify corectly what he need on server.

This solution work for me perfect 💯 

Hope your problems are solved also.
Cheers !

**Comment by PhilterPaper** [comment]
Another thing to check if you're getting `Blocked access to file` or other problems with local files (e.g., .js, .css, images), is to explicitly _enable_ local file access (`--enable-local-file-access`). The default changed as of 0.12.6 to refuse to access local files.

**Comment by LazyCompiler** [comment]
Still no official solution available, unbelievable 😞

I ended up doing:
```
try:
    pdfkit.from_file(path, new_path)
except OSError as e:
    if 'Done' not in getattr(e, 'message', str(e)):
        raise e
```

**Comment by LazyCompiler** [comment]
Still no official solution available, unbelievable 😞

Python's _pdfkit_ users? I ended up doing:
```
try:
    pdfkit.from_file(path, new_path)
except OSError as e:
    if 'Done' not in getattr(e, 'message', str(e)):
        raise e
```

**Comment by LazyCompiler** [comment]
Still no official solution available, unbelievable 😞

As for now, I solved it by doing that (Python pdfkit):
https://stackoverflow.com/a/67372025/12595469

**Comment by naveeduam** [comment]
any one having this issue in windows just make sure to use --enable-local-file-access in command line see below
wkhtmltopdf.exe --enable-local-file-access  filename.html  filename.pdf

**Comment by imsks** [comment]
I removed the SVGs files to resolve the issue because the library can't handle that.

**Comment by biskweet** [comment]
Still no fix for this error. Sad for a repository that has been updated less than 5 month ago.

## Workaround
The workaround is still (yes, even if you don't think that it isn't causing the problem) to transform all links to abolute links, whether they are local or online.
